test/conf/nginx.conf

```diff
@@ -165,6 +165,28 @@ http {
         }
     }
 
+    server {
+        listen      58072 ssl;
+        server_name  _;
+        ssl_protocols       TLSv1 TLSv1.1 TLSv1.2;
+        ssl_ciphers         HIGH:!aNULL:!MD5;
+        ssl_certificate     __NGXDOCROOT__/dummy.crt;
+        ssl_certificate_key __NGXDOCROOT__/dummy.key;
+        ssl_client_certificate __NGXDOCROOT__/dummy.crt;
+        ssl_verify_client on;
+
+        mruby_ssl_handshake_handler_code '
+            ssl = Nginx::SSL.new
+            ssl.certificate = "__NGXDOCROOT__/#{ssl.servername}.crt"
+            ssl.certificate_key = "__NGXDOCROOT__/#{ssl.servername}.key"
+            ssl.client_certificate = "__NGXDOCROOT__/#{ssl.servername}.crt"
```

Comment:
> 1. Wrong call back

Have you looked at X509_STORE_CTX_set_verify_cb()? 
A co-worker of mine suggests to look at it. Both of us are not openssl expert, though.

* https://linux.die.net/man/3/x509_store_ctx_set_verify_cb
* https://stackoverflow.com/questions/16291809/programmatically-verify-certificate-chain-using-openssl-api
* https://github.com/openssl/openssl/blob/master/apps/verify.c#L227
* https://github.com/openssl/openssl/blob/master/ssl/ssl_cert.c#L422-L423

Reply:
> 2. Not able to test the client cert with openssl s_client

Do you want to discuss #435 again or write openssl client with mruby?
Or any other ideas?

Reply:
> 3. Test failing for openssl version 1.0.2

Could you compare working version of build/nginx/logs/error.log and failed one?

We have some openssl errors with openssl 1.0.2. See https://travis-ci.org/matsumotory/ngx_mruby/jobs/639756579#L21570-L21607

```
2020/01/21 03:15:33 [debug] 9227#0: accept on 0.0.0.0:58072, ready: 0
2020/01/21 03:15:33 [debug] 9227#0: posix_memalign: 0000000002CD4B60:512 @16
2020/01/21 03:15:33 [debug] 9227#0: *102 accept: 127.0.0.1:53612 fd:29
2020/01/21 03:15:33 [debug] 9227#0: *102 event timer add: 29: 60000:564961
2020/01/21 03:15:33 [debug] 9227#0: *102 reusable connection: 1
2020/01/21 03:15:33 [debug] 9227#0: *102 epoll add event: fd:29 op:1 ev:80002001
2020/01/21 03:15:33 [debug] 9227#0: timer delta: 8
2020/01/21 03:15:33 [debug] 9227#0: worker cycle
2020/01/21 03:15:33 [debug] 9227#0: epoll timer: 60000
2020/01/21 03:15:33 [debug] 9227#0: epoll: fd:29 ev:0001 d:0000000002CE24B9
2020/01/21 03:15:33 [debug] 9227#0: *102 http check ssl handshake
2020/01/21 03:15:33 [debug] 9227#0: *102 http recv(): 1
2020/01/21 03:15:33 [debug] 9227#0: *102 https ssl handshake: 0x16
2020/01/21 03:15:33 [debug] 9227#0: *102 tcp_nodelay
2020/01/21 03:15:33 [debug] 9227#0: *102 SSL server name: "localhost"
2020/01/21 03:15:33 [debug] 9227#0: *102 ngx_mruby : mruby ssl handler: servername "localhost"
2020/01/21 03:15:33 [debug] 9227#0: *102 posix_memalign: 0000000002CD76E0:512 @16
2020/01/21 03:15:33 [debug] 9227#0: *102 ngx_mruby : mruby ssl handler: changing certificate to cert=/home/travis/build/matsumotory/ngx_mruby/build/nginx/html//localhost.crt key=/home/travis/build/matsumotory/ngx_mruby/build/nginx/html//localhost.key
2020/01/21 03:15:33 [debug] 9227#0: *102 ngx_mruby : mruby ssl handler: changing certificate to cert=/home/travis/build/matsumotory/ngx_mruby/build/nginx/html//localhost.crt
2020/01/21 03:15:33 [debug] 9227#0: *102 SSL_do_handshake: -1
2020/01/21 03:15:33 [debug] 9227#0: *102 SSL_get_error: 2
2020/01/21 03:15:33 [debug] 9227#0: *102 reusable connection: 0
2020/01/21 03:15:33 [debug] 9227#0: timer delta: 0
2020/01/21 03:15:33 [debug] 9227#0: worker cycle
2020/01/21 03:15:33 [debug] 9227#0: epoll timer: 60000
2020/01/21 03:15:33 [debug] 9227#0: epoll: fd:29 ev:0001 d:0000000002CE24B9
2020/01/21 03:15:33 [debug] 9227#0: *102 SSL handshake handler: 0
2020/01/21 03:15:33 [debug] 9227#0: *102 verify:1, error:0, depth:1, subject:"/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd", issuer:"/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd"
2020/01/21 03:15:33 [debug] 9227#0: *102 verify:0, error:7, depth:0, subject:"/C=JP/ST=Tokyo/L=Shibuya/O=Chain Corp/OU=Engineering/OU=Team/CN=John Doe/emailAddress=john@doe.com", issuer:"/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd"
2020/01/21 03:15:33 [debug] 9227#0: *102 verify:1, error:7, depth:0, subject:"/C=JP/ST=Tokyo/L=Shibuya/O=Chain Corp/OU=Engineering/OU=Team/CN=John Doe/emailAddress=john@doe.com", issuer:"/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd"
2020/01/21 03:15:33 [debug] 9227#0: *102 SSL_do_handshake: 1
2020/01/21 03:15:33 [debug] 9227#0: *102 SSL: TLSv1.2, cipher: "ECDHE-RSA-AES256-GCM-SHA384 TLSv1.2 Kx=ECDH Au=RSA Enc=AESGCM(256) Mac=AEAD"
2020/01/21 03:15:33 [debug] 9227#0: *102 reusable connection: 1
2020/01/21 03:15:33 [debug] 9227#0: *102 http wait request handler
2020/01/21 03:15:33 [debug] 9227#0: *102 malloc: 0000000002BBBDB0:1024
2020/01/21 03:15:33 [alert] 9227#0: *102 ignoring stale global SSL error (SSL: error:0407008A:rsa routines:RSA_padding_check_PKCS1_type_1:invalid padding error:04067072:rsa routines:RSA_EAY_PUBLIC_DECRYPT:padding check failed error:0D0C5006:asn1 encoding routines:ASN1_item_verify:EVP lib) while waiting for request, client: 127.0.0.1, server: 0.0.0.0:58072
2020/01/21 03:15:33 [debug] 9227#0: *102 SSL_read: -1
2020/01/21 03:15:33 [debug] 9227#0: *102 SSL_get_error: 2
```

Reply:
Thanks for your reply. 
1. I will try with your suggestion about call back.
2. I tried on my side with `openssl client`, but no luck. I would like to discuss again, to use C Ruby at least `openssl` test cases. But before that, let me make it work.
3. Thanks, I will try to find out with your suggestion.

Reply:
@yyamano 
I went throught some docs and I ended up with following conclusion.
1. About callback.
Thanks for your suggestion. But it wasn't match what I want to try. My idea was just switch CA certificates for Mutual authentication (client authentication) and leave verification for NGINX module (OpenSSL function). Because, customizing of verification might create lot of bugs. Even MRuby side we don't have proper OpenSSL gem. 
So, I think current callback should work fine.
2. About test
Thanks for your latest update. I'm using your lates update for my test.
3. No longer support 1.0.2 version. So, I think I can skip this issue.